### ISSUE TITLE
General question about coordinates and user input

### ISSUE THREAD

**Issue by Prismacolor** [pymupdf/PyMuPDF#439]
I attempted to send an email and I am unsure if it went through.  I have written a script based on this library which asks the user to put in their own start and stop points and the script will draw a rectangle based on those words. I was wondering about the coordinate structure, where does the library get the coordinates from? I ask this because I know there's a C# library which will allow you to identify the x and y coordinates of something and then extract text based on that, and I was wondering if this was similar. So theoretically, is it possible for a user to click on part of the pdf and us be able to retrieve the coordinates, or does it have to be a specific word? 

I also see there are some capabilities for working within the PDF is there capability to merge the PDFs all into one large document? I ask this because it's possible the rectangle my user needs to draw crosses over numerous pages and because I am not seeing the document before they upload it, I would not be able to accurately predict the height or width of the rectangle. 

Final question (I'm sorry there are so many) Is it possible to draw a large rectangle around your entire document, and then within that large rectangle identify smaller rectangles for smaller bits of text you might want? Thanks!

**Comment by JorjMcKie** [comment]
@Prismacolor - I think I received your e-mails, and I also answered them via the same channel. No idea why they did not reach you.
I am hesitant publishing the company name here, so let me simply copy my answers into this chat:

> PyMuPDF has no GUI component, which would offer ways to directly draw things like rectangles. On the other hand, it doesn’t care where the defining components of rectangles come from. It could be – as you indicate – keywords, or direct coordinates, page limits … whatever. Using a GUI, a user could also identify two points such that a rectangle could be constructed from them (using them as diagonally opposite points, or by just using their different height coordinates and otherwise taking the full page width as left / right delimiters …).
It would be helpful if e.g. you employee definition were somehow “encapsulated”, maybe a keyword as a starting point. And the end point could be another keyword, or the start keyword of the next employee data.
It is no problem to extract text from within an rectangle as such. Where appropriate, any image inside a rectangle can also be identified and extracted, etc. etc.

So basically, you are free to choose whatever way to pass rectangle information in to PyMuPDF.
I hope the above reflects what you actually are asking. Please let me know otherwise.

**Comment by Prismacolor** [comment]
Ah this is good okay, I guess my question is, if I ask my user to indicate let's say the starting point of a rectangle, does the user need to know what those coordinates are? Or do i have a way of finding out what the coordinates are of each  word on the document before hand so that when my user clicks I can grab the coordinates myself, and then draw a rectangle.  So far I have only ever typed in text and then had the coordinates returned to me.

**Comment by JorjMcKie** [comment]
I hope we are talking about the same things ...

1. There are several ways to extract the text of a page - among them again several which also deliver position information alongside the text itself. One out the many alternatives is ``Page.getText("words")`` which delivers a list of all words (i.e. strings without intervening spaces) together with rectangle coordinates encapsulating the word.
2. If you have a GUI script which displays a document page and then lets the user indicate an area on that page from which any text contained therein should be extracted, then you fairly easily identify and extract the respective text pieces.

**Comment by Prismacolor** [comment]
Okay got it, I think I misunderstood what you were saying earlier. I think this is enough to get me started, thank you so much for your help!

**Comment by Prismacolor** [comment]
sorry, did you also have any thoughts about the other two paragraphs in my original comment? My colleagues had asked me about making smaller rectangles, but I think then if i have the coordinates of the large rectangle I should be able to go in and just use some if else statements to restrict the coordinates of other rectangles.

**Comment by JorjMcKie** [comment]
your other questions can all be answered with **yes:**

1. you can freely join PDFs in any way you want, e.g. taking a few pages from number 1, then a few other pages from number 2, etc. Simultaneously you can revert the page sequence or rotate pages. The method ``Document.insertPDF()`` would be appropriate here.
2. handling of rectangles and other geometrical objects is also supported: joining and intersecting, enlarging rectangle by including one or more points, transform geometry objects with matrices (mirroring, rotating, shifting, shearing, ...). All that is also supported by using the usual mathematical algebra (+, -, *, /). E.g. ``rect += (-5, -5, 5, 5)`` enlarges rectangle **rect** by 5 pixels in every dimension and ``r = r1 & r2`` delivers the intersection of rectangles r1, r2.

>  Is it possible to draw a large rectangle around your entire document, and then within that large rectangle identify smaller rectangles for smaller bits of text you might want?

No, not out of the box. Wouldn't make much sense either. A document by its very nature consists of its **_pages_** (plus some page-independent metadata). Every page may also have a different page size. Any displayable content is displayed on one specific page, etc.
There is no imaginable general value in building a big rectangle covering all the pages - without then always calculating back and forth between this and the set of the single pages.

The text extraction methods deliver either strings, lists or dictionaries per each page. You can of course do what you want (combine / join / concatenate) with these page outputs by using appropriate Python means if you see any value there ...

**Comment by Prismacolor** [comment]
Ah okay so not the whole document, I got itt, but it seems like based on your response in No. 2 if I drew a large rectangle around some of the data on the page, it could be possible to draw smaller rectangles within that, using that larger rectangle to restrict the size of the smaller ones. We are trying to think of ways to organize/tag the data within that large rectangle.

**Comment by Prismacolor** [comment]
our PDFs unfortunately do not have a lot of useful meta data we can use to tag items within them.

**Comment by JorjMcKie** [comment]
> our PDFs unfortunately do not have a lot of useful meta data we can use to tag items within them.

This is not what I meant with the term "metadata".
Anyway, if you want and if your confidentiality conditions permit, you can send me a small (2 - 3 pages) example PDF, so I can have a look and better understand what you guys are trying to do.
Maybe I can provide a few hints on how to attack the problem.

**Comment by JorjMcKie** [comment]
> I also see there are some capabilities for working within the PDF is there capability to merge the PDFs all into one large document? I ask this because it's possible the rectangle my user needs to draw crosses over numerous pages and because I am not seeing the document before they upload it, I would not be able to accurately predict the height or width of the rectangle.

This can be solved in a fairly straightforward way. For example, 

* let your user identify a point where to start, e.g. p1 on page n1.
* then he identifies the end point p2 on page n2.
* then in your script extract text from page n1 which is inside the rectangle ``fitz.Rect(p1, width1, height1)`` (width1, height1 is the bottom-right point of page n1).
* then extract all text on pages with numbers n1 < n < n2.
* then extract text on page n2 within rectangle ``fitz.Rect(0, 0, p2)``.
* done

In order to have overall correct position info for all text extracted in this way, add the page heights from previous pages to the respective text bboxes. This will finally deliver an overall text corpus with the same position info as if extracted from a large rectangle.
Hope I made myself clear.

**Comment by JorjMcKie** [comment]
Assuming there are no more questions now.
